src/Associativity.cpp

```diff
@@ -0,0 +1,658 @@
+                         bool is_associative, vector<AssociativeOp> ops) {
+    auto result = prove_associativity(f, args, exprs);
+    internal_assert(result.first == is_associative)
+        << "Checking associativity: " << print_args(f, args, exprs) << "\n"
+        << "  Expect is_associative: " << is_associative << "\n"
+        << "  instead of " << is_associative << "\n";
+    if (is_associative) {
+        for (size_t i = 0; i < ops.size(); ++i) {
+            const AssociativeOp &op = result.second[i];
+            internal_assert(equal(op.identity, ops[i].identity))
+                << "Checking associativity: " << print_args(f, args, exprs) << "\n"
+                << "  Expect identity: " << ops[i].identity << "\n"
+                << "  instead of " << op.identity << "\n";
+            internal_assert(equal(op.x.second, ops[i].x.second))
+                << "Checking associativity: " << print_args(f, args, exprs) << "\n"
+                << "  Expect x: " << ops[i].x.second << "\n"
+                << "  instead of " << op.x.second << "\n";
+            internal_assert(equal(op.y.second, ops[i].y.second))
+                << "Checking associativity: " << print_args(f, args, exprs) << "\n"
+                << "  Expect y: " << ops[i].y.second << "\n"
+                << "  instead of " << op.y.second << "\n";
+        }
+    }
+
+    for (const auto &op : result.second) {
+        debug(4) << "Operator: " << op.op << "\n";
+        debug(4) << "   identity: " << op.identity << "\n";
+        debug(4) << "   x: " << op.x.first << " -> " << op.x.second << "\n";
+        debug(4) << "   y: " << op.y.first << " -> " << op.y.second << "\n";
+    }
+    debug(4) << "\n";
+}
+
+} // anonymous namespace
+
+void associativity_test() {
+    Expr x = Variable::make(Int(32), "x");
+    Expr y = Variable::make(Int(32), "y");
+    Expr z = Variable::make(Int(32), "z");
+    Expr rx = Variable::make(Int(32), "rx");
+
+    Expr f_call_0 = Call::make(Int(32), "f", {x}, Call::CallType::Halide, nullptr, 0);
+    Expr f_call_1 = Call::make(Int(32), "f", {x}, Call::CallType::Halide, nullptr, 1);
+    Expr f_call_2 = Call::make(Int(32), "f", {x}, Call::CallType::Halide, nullptr, 2);
+    Expr g_call = Call::make(Int(32), "g", {rx}, Call::CallType::Halide, nullptr, 0);
+
+
+    // f(x) = min(f(x), int16(z))
+    check_associativity("f", {x}, {min(f_call_0, y + Cast::make(Int(16), z))},
+                        true, {{Expr(), Int(32).max(), {"", f_call_0}, {"", y + Cast::make(Int(16), z)}}});
+
+    // f(x) = f(x) + g(rx) + y + z
+    check_associativity("f", {x}, {y + z + f_call_0},
+                        true, {{Expr(), make_const(Int(32), 0), {"", f_call_0}, {"", y + z}}});
+
+    // f(x) = max(y, f(x))
+    check_associativity("f", {x}, {max(y, f_call_0)},
+                        true, {{Expr(), Int(32).min(), {"", f_call_0}, {"", y}}});
+
+    // f(x) = Tuple(2, 3, f(x)[2] + z)
```

Comment:
Would be interested to see more Tuple tests, to get a sense of the limits in the multidimensional case. E.g, I imagine complex addition works, but would complex multiplication work?